warn/warn_test.go

```diff
@@ -114,3 +114,145 @@ func checkFindingsAndFix(t *testing.T, category, input, output string, expected
 		checkFix(t, category, output, output, scope, fileType)
 	}
 }
+
+func TestCalculateDifference(t *testing.T) {
+	tests := []struct {
+		before      string
+		after       string
+		start       int
+		end         int
+		replacement string
+	}{
+		{
+			before:      "asdf",
+			after:       "asxydf",
+			start:       2,
+			end:         2,
+			replacement: "xy",
+		},
+		{
+			before:      "asxydf",
+			after:       "asdf",
+			start:       2,
+			end:         4,
+			replacement: "",
+		},
+		{
+			before:      "asxydf",
+			after:       "asztdf",
+			start:       2,
+			end:         4,
+			replacement: "zt",
+		},
+		{
+			before:      "",
+			after:       "foobar",
+			start:       0,
+			end:         0,
+			replacement: "foobar",
+		},
+		{
+			before:      "foobar",
+			after:       "",
+			start:       0,
+			end:         6,
+			replacement: "",
+		},
+		{
+			before:      "qwerty",
+			after:       "asdfgh",
+			start:       0,
+			end:         6,
+			replacement: "asdfgh",
+		},
+		{
+			before:      "aa",
+			after:       "aaaa",
+			start:       2,
+			end:         2,
+			replacement: "aa",
+		},
+		{
+			before:      "aaaa",
+			after:       "aa",
+			start:       2,
+			end:         4,
+			replacement: "",
+		},
+		{
+			before:      "abc",
+			after:       "abdbc",
+			start:       2,
+			end:         2,
+			replacement: "db",
+		},
+		{
+			before:      "abdbc",
+			after:       "abc",
+			start:       2,
+			end:         4,
+			replacement: "",
+		},
+	}
+
+	for _, tc := range tests {
+		before := []byte(tc.before)
+		after := []byte(tc.after)
+
+		start, end, replacement := calculateDifference(&before, &after)
+
+		if start != tc.start || end != tc.end || replacement != tc.replacement {
+			t.Errorf("Wrong difference for %q and %q: want %d, %d, %q, got %d, %d, %q",
+				tc.before, tc.after, tc.start, tc.end, tc.replacement, start, end, replacement)
+		}
+	}
+}
+
+func TestSuggestions(t *testing.T) {
+	// Suggestions are not generated by individual warning functions but by the warnings framework.
+
+	contents := `foo()
+
+attr.bar(name = "bar", cfg = "data")
+
+attr.baz("baz", cfg = "data")
+`
+	f, err := build.ParseBzl("file.bzl", []byte(contents))
+	if err != nil {
+		t.Errorf("Parse error: %v", err)
+	}
+
+	findings := FileWarnings(f, "pkg", []string{"attr-cfg"}, nil, ModeSuggest)
+	want := []struct {
+		start       int
+		end         int
+		replacement string
+	}{
+		{
+			start:       28,
+			end:         42,
+			replacement: "",
+		},
+		{
+			start:       59,
+			end:         73,
+			replacement: "",
```

Comment:
why you only have tests without replacement?

Reply:
Because currently the only supported replacement is `attr-cfg` which removes `cfg = "data"`. Empty string doesn't mean there's no replacement, it means that the given range should be replaced with an empty string. I'll add more tests here once more warning functions support replacements including some that actually add something new to the file.